### DISCUSSION TITLE
ERR_CONNECTION_REFUSED when running Dusk in CI (Github Actions)

### DISCUSSION THREAD

**Discussion by DanielGSoftware** [laravel/framework#51404] [Q&A] [answered]
We've been experiencing a persistent issue with our dusk tests in CI for over a year. Approximately 95% of the time, a part of our tests fail with the error UnknownErrorException: unknown error: net::ERR_CONNECTION_REFUSED. Interestingly, the initial tests pass without any issues, but about halfway through, the remaining tests consistently encounter this error. Occasionally (about 5% of the time), all tests complete successfully without any errors. Despite several attempts to resolve the issue, we have not yet found a solution, which prevents use to use dusk tests in our CI process. Also, when this initial problem came up, there was no code related change that could've caused this, as the tests randomly started failing with no commit prior.

It seems that the issue appears less often when I reduce the number of Dusk tests in our application, although we only use about 9 tests files, which have 26 tests in total. The entire process takes around 5 minutes to complete on GitHub, but this duration is reduced because some tests fail quickly because of mentioned issue.

Here's what I've tried so far:
* Adjusting various configuration values in the environment (for session, cache, etc.)
* Switching our workflow to the [Laravel Dusk GH actions workflow](https://laravel.com/docs/11.x/dusk#running-tests-on-github-actions).
* Replacing our DuskTestCase with the original [DuskTestCase from Laravel Dusk](https://github.com/laravel/dusk/blob/8.x/tests/Browser/DuskTestCase.php).
* Experimenting with different browser configurations, such as --enable-file-cookies and --disable-dev-shm-usage.
* Adding and removing tests. ~~Interestingly, fewer tests sometimes seem to pass successfully, which is something I now realize could be significant and warrants further investigation.~~ after investigating, it didn't seem to have an impact.
* Various other changes that I've now forgotten, essentially any modifications that might influence the outcome.

I would appreciate any help or suggestions, as this issue has been quite frustrating and is impacting our QA capabilities for a while now. It's also interesting to me that I couldn't find anyone else struggling with this issue, so it would be really nice to figure this out. In any case, thanks for reading!

**Comment by DanielGSoftware** [comment]
Bump

**Reply by jameshulse** [reply]
We have the exact same issue in Github Actions.

> Experimenting with different browser configurations, such as --enable-file-cookies and --disable-dev-shm-usage.

Did you try `--no-sandbox`, ` --disable-setuid-sandbox` etc?

**Reply by lucaguindani** [reply]
Same here, did someone found a solution ?

**Reply by jameshulse** [reply]
No, I don't believe we have. Our dusk tests have been flaky enough that the team are starting to ignore them, so they aren't bringing us much value unfortunately.

**Reply by DanielGSoftware** [reply]
> Same here, did someone found a solution ?

Nope, unfortunately not.

@jameshulse sorry for the super-late reply and thank you for your suggestion, unfortunately trying `--no-sandbox`,  `--disable-setuid-sandbox` didn't resolve the issue for us.

**Reply by LegendEffects** [reply, marked as answer]
We also encountered this issue and lived with it for a few months but it was a reason to not add more dusk tests since it added to the failure rate.

The app is an API which is probably overwhelming the single threaded dev server. Our solution was to just use `octane:start` instead of `serve` but we can only do that since we use Octane, the only other way I can think of solving this would be to actually run your app with FPM as a service within the job.

**Reply by DanielGSoftware** [reply]
@LegendEffects amazing! We've installed octane as a dev dependency and use it for running our Dusk tests in CI, and it fixes the issue. Thanks a lot!
